cloudconfig/cloudinit/cloudinit_opensuse.go

```diff
@@ -0,0 +1,226 @@
+}
+
+// PackageMirror is defined on the PackageMirrorConfig interface.
+func (cfg *openSUSECloudConfig) PackageMirror() string {
+	mirror, _ := cfg.attrs["package_mirror"].(string)
+	return mirror
+}
+
+// AddPackageSource is defined on the PackageSourcesConfig interface.
+func (cfg *openSUSECloudConfig) AddPackageSource(src packaging.PackageSource) {
+	cfg.attrs["package_sources"] = append(cfg.PackageSources(), src)
+}
+
+// PackageSources is defined on the PackageSourcesConfig interface.
+func (cfg *openSUSECloudConfig) PackageSources() []packaging.PackageSource {
+	sources, _ := cfg.attrs["package_sources"].([]packaging.PackageSource)
+	return sources
+}
+
+// AddPackagePreferences is defined on the PackageSourcesConfig interface.
+func (cfg *openSUSECloudConfig) AddPackagePreferences(prefs packaging.PackagePreferences) {
+	// TODO (aznashwan): research a way of using yum-priorities in the
+	// context of a single package and implement the appropriate runcmds.
+}
+
+// PackagePreferences is defined on the PackageSourcesConfig interface.
+func (cfg *openSUSECloudConfig) PackagePreferences() []packaging.PackagePreferences {
+	// TODO (aznashwan): add this when priorities in yum make sense.
+	return []packaging.PackagePreferences{}
+}
+
+// Render is defined on the the Renderer interface.
+func (cfg *openSUSECloudConfig) RenderYAML() ([]byte, error) {
+	// Save the fields that we will modify
+	var oldruncmds []string
+	oldruncmds = copyStringSlice(cfg.RunCmds())
+
+	// check for package proxy setting and add commands:
+	var proxy string
+	if proxy = cfg.PackageProxy(); proxy != "" {
+		cfg.AddRunCmd(addPackageProxyCmd_opensuse(cfg, proxy))
+		cfg.UnsetPackageProxy()
+	}
+
+	// add appropriate commands for package sources configuration:
+	srcs := cfg.PackageSources()
+	for _, src := range srcs {
+		cfg.AddScripts(addPackageSourceCmds(cfg, src)...)
+	}
+	cfg.UnsetAttr("package_sources")
+
+	data, err := yaml.Marshal(cfg.attrs)
+	if err != nil {
+		return nil, err
+	}
+
+	// Restore the modified fields
+	cfg.SetPackageProxy(proxy)
+	cfg.SetAttr("package_sources", srcs)
+	if oldruncmds != nil {
+		cfg.SetAttr("runcmd", oldruncmds)
+	} else {
+		cfg.UnsetAttr("runcmd")
+	}
+
+	return append([]byte("#cloud-config\n"), data...), nil
+}
+
+func (cfg *openSUSECloudConfig) RenderScript() (string, error) {
+	return renderScriptCommon(cfg)
+}
+
+// AddCloudArchiveCloudTools is defined on the AdvancedPackagingConfig.
+func (cfg *openSUSECloudConfig) AddCloudArchiveCloudTools() {
+	src, pref := config.GetCloudArchiveSource(cfg.series)
+	cfg.AddPackageSource(src)
+	cfg.AddPackagePreferences(pref)
+}
+
+func (cfg *openSUSECloudConfig) getCommandsForAddingPackages() ([]string, error) {
+	var cmds []string
+
+	if newMirror := cfg.PackageMirror(); newMirror != "" {
+		cmds = append(cmds, LogProgressCmd("Changing package mirror does not yet work on OpenSUSE"))
+		// TODO(bogdanteleaga, aznashwan): This should work after a further PR
```

Comment:
Do these TODO statements still make sense? If they are for common code, should the code be shared between openSUSE and CentOS?